### PR TITLE
server: fix root disk size on vm reset

### LINKED ISSUE #3957: Root disk size reverts to its template size after rebuild
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Management Server
~~~

##### CLOUDSTACK VERSION

~~~
4.13
~~~

##### CONFIGURATION

CloudStack 4.13
vSphere 6.5

##### STEPS TO REPRODUCE

- resize root disk to larger size
- rebuild vm

##### EXPECTED RESULTS

CloudStack resize root disk to larger size after rebuild

##### ACTUAL RESULTS

- root disk size reverts to its original template size
- CloudStack reports root disk size has not been reverted

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

If VM details contain `rootdisksize`, volume entry in DB should reflect correct size when VM reset is performed.

Fixes #3957 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
code looks good @shwstppr but have you verified that the resize code is being called on a vm reset?

**Comment by shwstppr** [comment]
@DaanHoogland I stopped work on this after your comment on the issue and reconsidering reset behaviour.
I think to support not reverting root disk size to the original value, maybe we can have a new param in resetVirtualMachine API. Default behaviour will revert size but user can call the API with the new flag to not revert it. However, that may require additional changes and testing.

**Comment by DaanHoogland** [comment]
@shwstppr (cc @Spaceman1984 ) the consensus so far is to revert to the template including the template root size disk. making it optional is outside of this scope.

**Comment by rhtyd** [comment]
cc @Pearl1594 @harikrishna-patnala @nvazquez

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by DaanHoogland — APPROVED** [review]
looks good, doing a new build/smoke test run on kvm

**Review by Pearl1594 — APPROVED** [review]
Verified behavior. LGTM
1. Created a VM with a template of size 3GB but overrode root disk size to 8GB. Reinstalled the VM. The re-installed VM's root disk size = 8GB
2. Created a VM with a template of size 3GB but overrode root disk size to 8GB. Performed a volume resize operation - volume size (root) = 15GB. Reinstalled VM - the reinstalled VM's root volume size = 15GB
3. Created a VM with a template of size 3GB. Resized the volume of the VM to 8GB. Reinstalled the VM. Re-installed VM's root disk size = 8GB

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -6678,6 +6678,12 @@ public UserVm restoreVirtualMachine(final Account caller, final long vmId, final
                 newVol = volumeMgr.allocateDuplicateVolume(root, null);
             }
 
+            if (userVmDetailsDao.findDetail(vm.getId(), VmDetailConstants.ROOT_DISK_SIZE) == null && !newVol.getSize().equals(template.getSize())) {
+                VolumeVO resizedVolume = (VolumeVO) newVol;
+                resizedVolume.setSize(template.getSize());
+                _volsDao.update(resizedVolume.getId(), resizedVolume);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6684]: @shwstppr it is possible to handle this case while allocating duplicate volume with template ?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -6678,6 +6678,12 @@ public UserVm restoreVirtualMachine(final Account caller, final long vmId, final
...
+            }
+
             // 1. Save usage event and update resource count for user vm volumes
             _resourceLimitMgr.incrementResourceCount(newVol.getAccountId(), ResourceType.volume, newVol.isDisplay());
             _resourceLimitMgr.incrementResourceCount(newVol.getAccountId(), ResourceType.primary_storage, newVol.isDisplay(), new Long(newVol.getSize()));
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6689]: `newVol` here doesn't have updated template size, refresh it after the template size is updated in DB.

Comment (shwstppr) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6689]: @sureshanaparti same object is been cast to VolumeVO and size is updated  at 6682-6683. `newVol` will return the updated size

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 1 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6684]: @sureshanaparti Duplicating using template, won't that affect other existing values of the entry - folder, path, etc?

**Comment by rhtyd** [comment]
Merged based on Daan, Pearl's review testing and smoketests

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 1 above)
Comment (rhtyd) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:6684]: I missed this comment, if we need to address anything additional pl raise a new PR @shwstppr @sureshanaparti

**Pull request merged**